rfd/xxxx-command.md

```diff
@@ -0,0 +1,208 @@
+```
+Teleport Proxy will be used as the main command execution engine.
+
+* It will allow parallel execution of commands on multiple nodes (configurable execution pool size).
+* It will allow to stream output from multiple nodes to the client.
+* Nodes will record the session and upload in the same way as interactive session are recorded.
+* It will ensure that even in case of a network failure the command will continue to run on the node.
+  (all commands should be one shot. If an initial connection to the node fails, the proxy will retry.
+  If the connection is lost during the command execution, the command will continue to run on the node
+  and proxy will not re-try it as we have no guarantee that the command is idempotent).
+* (Optional) It will allow to specify a timeout for the command execution. If the command execution takes longer
+  than the specified timeout, the command will be terminated on the node and the output will be streamed
+  to the client.
+
+### Command resource
+
+We will introduce a new resource type `command` that will allow to define
+commands that can be executed by Teleport Command. The resource will have
+the following fields:
+
+```yaml
+#
+# Example resource for a Command
+#
+kind: command
+version: v1
+metadata:
+  # The name of the command. It must be unique.
+  name: cpu-usage
+  # Human-readable description
+  description: Show top 10 CPU usage
+spec:
+  # Interpreter to use for command execution. If not specified, /bin/sh will be used.
+  # This will allow to execute short Python or JS scripts.
+  interpreter: "/bin/bash"
+  command: |
+    ps -eo pid,ppid,cmd,%mem,%cpu --sort=-%cpu | head
+```
+
+Interpreter is an optional field. It should allow specifying the interpreter
+allowing to not only execute a shell command but also a short Python or JS script.
+
+#### Picking up the OS user
+
+The user will be provided as a command argument. If the user doesn't exist on the node,
+the command execution will fail. If a user want to execute a command as two different
+users, then one command should be execute for each user.
+An alternative would be to add the user to `ssh_service` configuration, but this
+will require modifying the configuration on all nodes. 
+
+## Roles
+
+Roles will be extended to allow defining access to commands. The following
+example will allow to execute commands with `environment=dev-cmd` label on nodes
+with `environment=dev` label. It will also deny access to commands with
+`environment=prod` label.
+This approach will allow to define a set of commands that can be executed
+on a particular set of nodes (e.g. `dev-cmd` commands can be executed on `dev` nodes
+but not on `prod` nodes).
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: developer
+spec:
+  allow:
+    # Labels selector for command execution.
+    command_execution_labels:
+      environment: ["dev-cmd"]
+    # Labels selector for commands resource access.
+    command_labels:
+     environment: ["dev"]
```

Comment:
I wonder if it will get too confusing, having 2 sets of new label selectors. Is there a reason we wouldn't want to keep using existing "node_labels" field and e.g. have a role option that controls whether a user is allowed interactive shell or not? Something like

```yaml
spec:
  options:
    allow_interactive_shell: false // indicates only can use Commands
  allow:
    node_labels:
      env: ["dev"]
    command_labels:
      env: ["dev"]
```

WDYT?

Reply:
1. I think that `allow_interactive_shell` is confusing. It suggests that `tsh ssh nodeName ls -la` is still allowed, as this is non-interactive execution, but this won't be the case.
2. IMO this is more confusing. Having new labels `command_labels` and `command_execution_labels` directly refer to the new feature. Adding a boolean option creates this weird dependency. If `node_labels` are specified and `allow_interactive_shell` is set to `true` then the user has access to the interactive shell, but if `allow_interactive_shell` is set to `false` then `node_labels` refers to Teleport Command, not SSH access anymore.

Reply:
1. I think that `allow_interactive_shell` is confusing. It suggests that `tsh ssh nodeName ls -la` is still allowed, as this is non-interactive execution, but this won't be the case.
2. IMO this is more confusing. Having new labels `command_labels` and `command_execution_labels` directly refer to the new feature. Adding a boolean option creates this weird dependency. If `node_labels` are specified and `allow_interactive_shell` is set to `true`, then the user has access to the interactive shell, but if `allow_interactive_shell` is set to `false`, then `node_labels` refers to Teleport Command, not SSH access anymore.

But I agree, the proposed solution is not very intuitive. I just don't have a better idea of how we can implement it in a different way.

Reply:
> the proposed solution is not very intuitive. I just don't have a better idea of how we can implement it in a different way.

The access request system has an established pattern of always using a reference to an external role as the model for privilege escalation.  E.g. when requesting expanded access generally you literally request additional roles, and for per-resource access requests which resources you're allowed to request access to is defined by your `search_as_roles` directive.  Reasonable people can disagree about how _intuitive_ that is, but it is consistent, and it has the big upside of not needing to maintain multiple separate implementations for what constitutes an RBAC selector for a given resource (e.g. the access request system will benefit from the new expression-based rbac rules automatically without needing to change any access-request specific code).

Could do something similar with commands.  Ex:

```yaml
command_labels:
  env: prod
command_as_roles: ["prod-access"]
```